vaadin-grid-flow-demo/src/main/java/com/vaadin/flow/component/grid/demo/GridDemo.java

```diff
@@ -1675,6 +1651,67 @@ private void createItemDetailsOpenedProgrammatically() {
         addCard("Item details", "Open details programmatically", header, grid);
     }
 
+    // begin-source-example
+    // source-example-heading: Navigating grid items externally
+    // Clicking on item in grid opens dialog for item navigation
+    private Grid createExternalDataNavigationGrid() {
+        Grid<Person> grid = new Grid<>(Person.class);
+        final GridListDataView<Person> dataView = grid
+                .setDataProvider(new PersonService().fetchAll());
+
+        grid.removeColumnByKey("id");
+
+        // The Grid<>(Person.class) sorts the properties and in order to
+        // reorder the properties we use the 'setColumns' method.
+        grid.setColumns("firstName", "lastName", "age", "address",
+                "phoneNumber");
+
+        grid.addItemClickListener(
+                event -> new DataDialog(dataView, event.getItem()).open());
+
+        return grid;
+    }
+
+    private class DataDialog extends Dialog {
+        private Button next;
+        private Button previous;
+        private Person currentItem;
+        private final GridListDataView<Person> dataView;
+        private Span data = new Span();
+
+        public DataDialog(GridListDataView<Person> dataView, Person item) {
+            this.dataView = dataView;
+            next = new Button("Next",
+                    event -> setItem(dataView.getNextItem(currentItem)));
+            previous = new Button("Previous",
+                    event -> setItem(dataView.getPreviousItem(currentItem)));
+            setItem(item);
+            setModal(true);
+            HorizontalLayout layout = new HorizontalLayout(previous, data, next);
+            layout.expand(data);
+            layout.setAlignItems(FlexComponent.Alignment.CENTER);
+            layout.setWidth("400px");
+            add(new VerticalLayout(new Span("Click outside to close"), layout));
+        }
+
+        private void setItem(Person item) {
+            currentItem = item;
+            data.setText(String.format("%s %s", item.getFirstName(), item.getLastName()));
+            next.setEnabled(dataView.hasNextItem(currentItem));
+            previous.setEnabled(dataView.hasPreviousItem(currentItem));
+        }
+    }
+    // end-source-example
+
+    private void createExternalDataNavigation() {
+        Grid grid = createExternalDataNavigationGrid();
```

Comment:
I don't understand why are these in their own method ?

Reply:
It's to get the full source example without the `addCard(...)`. It's not the same as for the more complex Filter example.